File: ADT/stack-queue-pblm/739.py
# Google, Meta
# Monotonic Stack => maintaining a order (increasing or decreasing order)
# In Stack, Each element only push once or pop once, never push or pop more than one, only twice works push + pop => O(n * 2) => O(n)

class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        n = len(temperatures)
        st = []
        result = [0] * n
        for i in range(n - 1, -1, -1):
            while st and temperatures[i] >= temperatures[st[-1]]:
                st.pop()
            if not st:
                result[i] = 0
            else:
                result[i] = st[-1] - i
            st.append(i)
        return result

File: ADT/stack-queue-pblm/739.js
/**
 * @param {number[]} temperatures
 * @return {number[]}
 */
var dailyTemperatures = function (temperatures) {
    const n = temperatures.length
    const st = []
    const result = new Array(n)
    for (let i = n - 1; i >= 0; i--) {
        while (st.length && temperatures[i] >= temperatures[st[st.length - 1]]) {
            st.pop()
        }
        if (st.length === 0) {
            result[i] = 0
        } else {
            result[i] = st[st.length - 1] - i
        }
        st.push(i)
    }
    return result
};